### DISCUSSION TITLE
Random Ramblings

### DISCUSSION THREAD

**Discussion by HarlemSquirrel** [keeweb/keeweb#2147] [General Discussion]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by HarlemSquirrel** [comment]
@Aetherinox @antelle I enabled discussions so we don't have to keep chatting on that PR 😁

**Reply by Aetherinox** [reply]
Nice. Sorry, right now I've been fighting with Google Cloud Services. The API key for the local environment kept coming back that it was invalid; so I had to register an app so that I could get an API key that worked.

Now I'm looking into Microsoft Teams because that seems to be a large issue as well.

**Reply by HarlemSquirrel** [reply]
Thanks for tackling these!

**Reply by Aetherinox** [reply]
No problem, I enjoy it.

This Microsoft Teams one is a pain. I don't want to bother antelle and get the proper configuration, but I'm currently looking for the correct redirectURL required by Microsoft Azure. Since this is a desktop app, there should be no need to redirect to the keeweb.info website. Not unless it's supposed to be `https://localhost`.

I deal with Google's Cloud portal, I've never messed with an Azure app. So it's making my brain hurt.

**Edit**: My comment gave me an idea, to try localhost, and that got me somewhere
```
http://localhost:8085/oauth-result/msteams.html
```

But that begs the question; what happens if the user changes the port? Or what about running a docker container and it using port 443.

**Edit 2**: Ok, well luckily the Microsoft documentation states
> Due to ephemeral port ranges often required by native applications, the port component (for example, :5001 or :443) is ignored for the purposes of matching a redirect URI. As a result, all of these URIs are considered equivalent: 

So that solves that issue, but the other issue is the official keeweb.info website. You're not allowed to have dissimilar TLDs, so if I use localhost as the TLD, all of them have to use localhost. If I put keeweb.info in there, it throws an error about it not matching the other results.

But some of their docs make no sense, because it says that using `127.0.0.1` is preferred, but if I use that, MS complains that it's not a valid address. It also says wildcards are allowed, but then the validation checker says "Do not use wildcards".

**Reply by antelle** [reply]
For Dropbox we have:
![IMG_7502](https://github.com/keeweb/keeweb/assets/633557/c753a90e-308a-433e-9582-8c1c39774421)

For teams we have three different apps: [local, prod, desktop](https://github.com/keeweb/keeweb/blob/master/app/scripts/const/cloud-storage-apps.js#L27), with different redirect URI’s.

**Comment by HarlemSquirrel** [comment]
GitHub also has a merge queue feature. I tied it on another project but it doesn't have high enough traffic to really see the benefits.

**Reply by Aetherinox** [reply]
@HarlemSquirrel I'd like to take a look at the branch for v1.18 and see what the diff is between that and master and what features would be pushed. Since it has been awhile, making sure there's nothing major.

Then get that branch caught up with master, and then start working on an actual v1.18 / 19 release.

There's a few bugs I've fixed, as well as a minor adjustment to otpauth URLs, and I'd like to crack out some more issues and start getting bugs patched for a next release.

One major task is getting all of the integration methods to function / making sure they all work. There's a ticket related to Dropbox integration not working, so I want to test it out and see if it's a user issue, or if something really is going on. The MSTeams one is a pain because I'm trying to cram years of documentation into my brain in a few days. Plus I think because I don't have a "paid" teams account, there's stuff I'm not seeing that I think I should, such as "Files". Teams is just a mess in general; there's 3 damn versions of the program, one has been removed (Classic), and now there's a Classic Free which is a severely stripped down version of Teams, and then a Teams paid version, and then a newer Teams version which is only for Work / School, and personal accounts cannot use it.

**Reply by HarlemSquirrel** [reply]
I have been using Dropbox for years with Kee web and it's worked flawlessly. What's the issue? I can look into that.

I also started trying to get builds working with node 18. Got most of the way there but maybe that's not the best place to spend time yet.

**Reply by Aetherinox** [reply]
Well, here's the short version. MSTeams integration has been reported to not work for quite a few people.

If you view the file `\app\scripts\storage\impl\storage-teams.js`, there is this line:

```
let scope = 'Sites.ReadWrite.All Team.ReadBasic.All';
if (!this.appSettings.shortLivedStorageToken) {
    scope += ' offline_access';
}
```

These are the scopes that are needed for msteams to work properly.

I created a new client id / secret through https://portal.azure.com/ and plugged the values into: `app\scripts\const\cloud-storage-apps.js`, however, if you open the developer console and then click Microsoft Teams and sign in, it returns an oauth error:

```
invalid_scope The provided value for the input parameter 'scope' is not valid. The scope 'Sites.ReadWrite.All Team.ReadBasic.All offline_access' does not exist.
```

Even though in Azure, I I have set the proper permissions:

![GN0Tbzs](https://github.com/keeweb/keeweb/assets/118329232/500d8088-81df-4ebd-bdb3-6c0c7bb11f93)

After doing some reading, someone recommended that Microsoft may have changed the way permissions are done, and to try:

```
let scope = 'https://graph.microsoft.com/.default';
```

And then I try to sign in to Microsoft Teams again. That time in the developer console I actually get much more of a response, but it's a 403

```
2024-04-07T01:55:38.493Z [storage-msteams] HTTP request POST https://login.microsoftonline.com/common/oauth2/v2.0/token
logger.js:43 2024-04-07T01:55:39.371Z [storage-msteams] HTTP response 200
logger.js:35 2024-04-07T01:55:39.371Z [storage-msteams] OAuth code exchanged {token_type: 'Bearer', scope: 'https://graph.microsoft.com/.default https://graph.microsoft.com/User.Read', expires_in: 3600, ext_expires_in: 3600, access_token: 'EwBgA8l6BAAUTTy6dbu0OLf3Lzl3R/vfUXxq8g8AAUQZpRZK8o…dsmO1IVpf6gLXyqbhIP/+d+LFR0d1WQFVeEnZ/rNSroeTg3cC', …}
logger.js:35 2024-04-07T01:55:39.372Z [storage-msteams] OAuth token received
logger.js:35 2024-04-07T01:55:39.372Z [storage-msteams] List undefined
logger.js:43 2024-04-07T01:55:39.372Z [storage-msteams] HTTP request GET https://graph.microsoft.com/v1.0/me/joinedTeams
storage-base.js:139 
        
  
       GET https://graph.microsoft.com/v1.0/me/joinedTeams 403 (Forbidden)
_httpRequestWeb @ storage-base.js:139
Show 1 more frame
Show less
logger.js:43 2024-04-07T01:55:40.062Z [storage-msteams] HTTP response 403
logger.js:59 2024-04-07T01:55:40.063Z [storage-msteams] List error 691ms http status 403
```

Then if you attempt to sign into Microsoft Teams again, you get:

![chUKUzj](https://github.com/keeweb/keeweb/assets/118329232/03d47580-fa1e-4bce-ac08-72294a122ac5)

I feel like I'm "sort of" heading in the right direction, but as I said, Azure is new to me. I do Google Cloud, and I can set up an app in less than 5 minutes, but this system makes zero sense. Because all of the results I'm finding online, one group of people give one set of instructions, and then another group gives entirely different instructions.

I can confirm however, that the current implementation of Microsoft Teams is broke. Doesn't work on web, or Desktop.

I even used the instructions here for reference: https://github.com/keeweb/keeweb/wiki/Dropbox%20and%20GDrive But for Azure I did msteams instead of onedrive.

But this implementation is over 3 years old, so maybe the way scopes are called is different now. But finding answers is a pain when every single search result for Azure pops up with the Microsoft Community forums; and personally, I hate those replies, they are canned and robotic., and almost never target the specific issue. They're just generic answers.

**Reply by HarlemSquirrel** [reply]
Ugh. I've set up OAuth and SAML on several platforms but if there's no official documentation on this that may mean it's not really supported and anything we get working could break at any time

**Reply by Aetherinox** [reply]
I can already tell that me and Prettify are going to have issues.

I have a certain style I'm used to writing code, and prettify blocks me from doing anything. I had to switch over to pushing on my dev repo, because I'm six commits in on the same exact code, and it's complaining about spaces. Had to go install eslint for vscode.

**Reply by HarlemSquirrel** [reply]
I added a command to auto fix lints on my node 18 branch

https://github.com/keeweb/keeweb/pull/2149/commits/1c5adfc8c5ea580ef4ad0186d59bc64509ffcf9b

**Reply by Aetherinox** [reply]
It's a little easier now that I have eslinter in Visual Studio Code. I don't have to guess what the heck it means.

Bigger update incoming sometime soon. Fixing bugs, adding a few new features.

![h0P3NkE](https://github.com/keeweb/keeweb/assets/118329232/ee1c69c5-1ed2-41b2-b8ca-9ce8492c92cd)

Working on passphrase implementation as well like how KeePassXC has it.

Adding "Spaces" was an interesting task.

![ml0FsPJ](https://github.com/keeweb/keeweb/assets/118329232/3a15504d-d19a-4116-8dcb-31f102d7c3e7)

Because if you just add a space within the "include" values; it throws spaces for like every other letter, depending on the number of character types you have. So I had to do some logic so that it's a little smart about how spaces are distributed. Plus ensuring that spaces are distributed randomly, and there's no pattern to how they're added.

There's a mountain of stuff to modify / update / etc. But I wanted to at least add a few minor things to get development going again so users have at least something fresh until we can do things like migrate to react.

![C4SBrSm](https://github.com/keeweb/keeweb/assets/118329232/c81e44b7-7d1d-4665-93c4-2bf70eb932a3)

Passphrases needs some adjustments, because you should be able to change the length without it resetting back to a normal password.

There's a lot of settings I can do, but the interface needs to be re-worked a bit to allow for such settings.

**Reply by Aetherinox** [reply]
We're making progress

![xNHcC17](https://github.com/keeweb/keeweb/assets/118329232/426630f1-54dc-4c02-86cd-a86c479b61b5)

**Comment by clicky6** [comment]
I have different autotype settings for different entries so it would be helpful if near auto-type when right clicking it could show auto-type and then [combination] for example 
![Screenshot 2024-11-20 at 7 52 31 PM](https://github.com/user-attachments/assets/4f8d351e-f6c0-461e-aba3-1f989f5f98bb)

Also it would be great if in autotype it could have many options like autotype username autotype username enter autotype password autotype password enter I know I can just do that by double clicking on entry but there is no enter options with them

So are there any issues open with that, Should I open an issue?

**Comment by clicky6** [comment]
Also for desktop apps to restrict screen capture or screenshot would be great so that no app or anything can see it except ourselves
